### PR TITLE
add token service support for docker registry

### LINKED ISSUE #2316: Integrate or Interface Docker Registry
I know, this may be completely out of scope for this project. But I still try:

It would be cool if gitea could provide support for a docker registry.

Full integration is probably very unlikely to ever being added. But it would already help if we could interface with a standalone registry running in some container. The core features should be:

 * Authentication against shared user base
 * Authorization with shared repository permissions
 * Simple instructions for how to set up the above

### PR DESCRIPTION
**Pull request by a1012112796**
Fix #2316

- repo name ruler
  ``<owner>/<repo_name>/<image_name>``
- add a new repo permission unit named as `package`, it's not a
  real unit in gitea, just use for permission check.
- add a small api `GET /repos/{owner}/{repo}/packages/{type}/{name}` for
  test, maybe can add more later

ref:
- https://docs.docker.com/registry/spec/auth/token/
- https://docs.docker.com/registry/notifications/
- https://github.com/cesanta/docker_auth

TODO:
~~- [ ]  handle delete event~~ (can't real delet a repo in docker registry, so this work is not necessary ...)
- [ ] test code

### CODE REVIEW

File: routers/repo/package.go [hunk 1]
```diff
@@ -0,0 +1,421 @@
...
+	"code.gitea.io/gitea/modules/context"
+	"code.gitea.io/gitea/modules/log"
+	"code.gitea.io/gitea/modules/setting"
+	"code.gitea.io/gitea/modules/timeutil"
+
+	"github.com/docker/distribution/notifications"
+	"github.com/docker/distribution/registry/auth/token"
+)
+
+// DockerPluginLogin tocken service for docker registry
```
Comment (techknowlogick) [inline routers/repo/package.go]:
```suggestion
// DockerPluginLogin token service for docker registry
```

File: routers/repo/package.go [hunk 2]
```diff
@@ -0,0 +1,421 @@
...
+		return
+	}
+
+	token := ctx.Req.Header.Get("X-Token")
+	if token != setting.Docker.NotifyToken {
+		ctx.Error(http.StatusUnauthorized)
+		return
+	}
+
+	body, err := ioutil.ReadAll(ctx.Req.Body)
```
Comment (techknowlogick) [inline routers/repo/package.go]: `ioutil.ReadAll` is concerning, there should be a limit here

File: modules/structs/package.go [hunk 3]
```diff
@@ -0,0 +1,27 @@
+// Copyright 2021 The Gogs Authors. All rights reserved.
```
Comment (lunny) [inline modules/structs/package.go]: Gogs?

File: options/locale/locale_en-US.ini [hunk 4]
```diff
@@ -1957,6 +1957,9 @@ topic.done = Done
 topic.count_prompt = You can not select more than 25 topics
 topic.format_prompt = Topics must start with a letter or number, can include dashes ('-') and can be up to 35 characters long.
 
+packages = packages
+packages.desc = permission unit for third part packages
```
Comment (wULLSnpAXbWZGYDYyhWTKKspEQoaYxXyhoisqHf) [inline options/locale/locale_en-US.ini]: I don't mean to be rude but it should be third *party*.
feel free to tell me if spell-checking is not welcome. 🙈

File: models/unit.go [hunk 5]
```diff
@@ -255,6 +260,14 @@ var (
 		5,
 	}
 
+	UnitPacages = Unit{
```
Comment (wULLSnpAXbWZGYDYyhWTKKspEQoaYxXyhoisqHf) [inline models/unit.go]: should be `UnitPacages` --> `UnitPackages`?
same for L281 below, then.

File: models/package.go [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+			RepoID: repoID,
+			Name:   name,
+			Type:   typ,
+		}
+	}
+
+	return pkg, nil
+}
+
+// UpdateCols updates some columns
```
Comment (wULLSnpAXbWZGYDYyhWTKKspEQoaYxXyhoisqHf) [inline models/package.go]: Could I ask what columns this function updates?
Some package-related stuff in the db, I'd guess?
Pardon me if it's self-evident, I am not really a go programmer.

File: models/package.go [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+		Name:   option.Name,
+		Type:   option.Type,
+		RepoID: option.Repo.ID,
+	}
+	pkg.LowerName = strings.ToLower(pkg.Name)
+	_, err := x.Insert(pkg)
+	return err
+}
+
+// GetPackage get an package
```
Comment (wULLSnpAXbWZGYDYyhWTKKspEQoaYxXyhoisqHf) [inline models/package.go]: `get an package` --> `get a package`

**Comment by a1012112796** [comment]
@wULLSnpAXbWZGYDYyhWTKKspEQoaYxXyhoisqHf Thanks for your carefully check

**Review by wULLSnpAXbWZGYDYyhWTKKspEQoaYxXyhoisqHf — APPROVED** [review]
lgtm 🚀

File: models/package.go (hunk 6 above)
Comment (wULLSnpAXbWZGYDYyhWTKKspEQoaYxXyhoisqHf) [reply models/package.go]: Alright, I see.
Thanks!

File: modules/packages/docker/event_hook.go [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+					return err
+				}
+				if pkg != nil {
+					pkgs[event.Target.Repository] = pkg
+				}
+			}
+
+			// update update time
+			pkg.UpdatedUnix = timeutil.TimeStamp(event.Timestamp.Unix())
+			err = pkg.UpdateCols("updated_unix")
```
Comment (lafriks) [inline modules/packages/docker/event_hook.go]: This should probably use it's own function in model like `UpdateLastUpdated` or something like that than function like `UpdateCols` would not be needed

**Comment by lafriks** [comment]
I was thinking on maybe this could be somehow refactored that this configuration is in database so that it could be extended so that we could have server-wide or per-organization/user docker registry server so that for example I could host my own docker registry server but connect it to external gitea service (like gitea.com or codeberg.org) just for my organization

**Comment by techknowlogick** [comment]
@lafriks you could do exactly that currently with https://github.com/cesanta/docker_auth/ (I am the maintainer of that project)

**Comment by lunny** [comment]
An alternative to external docker registry integration is necessary because the primary aim of Gitea is to be a package store.

**Comment by lafriks** [comment]
> @lafriks you could do exactly that currently with https://github.com/cesanta/docker_auth/ (I am the maintainer of that project)

How this would allow integration with gitea auth?

**Comment by a1012112796** [comment]
> > @lafriks you could do exactly that currently with https://github.com/cesanta/docker_auth/ (I am the maintainer of that project)
> 
> How this would allow integration with gitea auth?

I know it's steps, It's a littile long.

steps:
- 1. user login on ``doecker auth service`` by the login link
-  2. ``doecker auth service`` try to request auth on target sevice (github,gitea...) follow oauth2 ruler
-  3. if login success,  ``doecker auth service`` will generate a token and response it to user
-  4. user request ``JWT token`` that docker reqistry need by this token from   ``doecker auth service``. done.

but user can't used password or api token on gitea or github to request ``JWT token`` directly from 
``doecker auth service`` ...

**Pull request closed without merging**